pkg/languagedetection/internal/detectors/go_detector.go

```diff
@@ -0,0 +1,45 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux
+
+package detectors
+
+import (
+	"debug/elf"
+	"fmt"
+
+	dderrors "github.com/DataDog/datadog-agent/pkg/errors"
+	"github.com/DataDog/datadog-agent/pkg/languagedetection/languagemodels"
+	"github.com/DataDog/datadog-agent/pkg/network/go/binversion"
+)
+
+type GoDetector struct{}
+
+// DetectLanguage allows for detecting if a process is a go process, and it's version.
+// Note that currently the GoDetector only returns non-retriable errors since in all cases we will not be able to detect the language.
+// Scenarios in which we can return an error:
+//   - Invalid permissions. The system-probe is not running as root, or the container does not have CAP_PTRACE.
+//   - Program exits early, and we fail to call `elf.Open`. Note that in the future it may be possible to lock the directory using a system call.
+//   - Program is not a go binary, or has build tags stripped out. In this case we return a `dderrors.NotFound`.
+func (GoDetector) DetectLanguage(pid int) (languagemodels.Language, error) {
+	exePath := getExePath(pid)
+
+	bin, err := elf.Open(exePath)
+	if err != nil {
+		return languagemodels.Language{}, fmt.Errorf("open: %v", err)
+	}
+	defer bin.Close()
```

Comment:
We should consider to resolve the path into the full container path, and cache the results. 
That way we'll reduce the CPU and memory consumption for parsing the same binary multiple times.
The most common use case is `runc` binary which runs a lot in k8s context, each time the PID will be different but the binary itself remain the same.

Currently I'm testing such an improvement (for the block-list purposes in USM)
https://github.com/DataDog/datadog-agent/pull/18420

Reply:
I think that's a really great idea, however I think it might require some additional testing. Do you think it might be a good idea to include this in a follow up, or is this a blocker in your opinion?

Reply:
that can be in a followup PR for sure